[docs] [migrate] updates to the performance page (#15688)
yugabyte/yugabyte-db: docs/content/preview/migrate/performance.md

  {{< note title="Note" >}}

yb-voyager only disables all the constraints except check constraints on the column in internal sessions to migrate data.

  {{< /note >}}

- Disables triggers during import in the target database.

### Techniques to improve performance

Use one or more of the following techniques to improve import performance:

- **Load data in parallel**. yb-voyager executes N parallel batch ingestion jobs at any given time, where N is equal to half of the total number of cores in the YugabyteDB cluster. Normally this is considered good default practice. However, if the target cluster runs on high resource machines with a large number of CPU cores, the default may result in underusing CPU resources.

  Use the [-–parallel-jobs](../yb-voyager-cli/#parallel-jobs) argument with the import data command to override the default setting. Set `--parallel-jobs` to a number higher than the default based on your cluster requirement.

>   If CPU use is greater than 50-60%, you should lower the number of jobs. Similarly, if CPU use is low, you can increase the number of jobs.

Reviewer: Add a note after this --parallel-jobs point 
```
Note: 
	 If timeouts are witnessed then restarts should be tried with less parallel-jobs.
```

> - **Increase batch size**. If the [--batch-size](../yb-voyager-cli/#batch-size) (default is 100000) is too small, the import will run slower because the time spent importing data may be comparable or less than the time spent on other tasks, such as bookkeeping, setting up the client connection, and so on. Increasing the batch-size to a very high value is not recommended as the whole batch is executed in one transaction.

Reviewer: `(default is 100000)`, default batch-size is 20000, please check in docs I mentioned it there.

- **Add disks** to reduce disk write contention. YugabyteDB servers can be configured with one or multiple disk volumes to store tablet data. If all tablets are writing to a single disk, write contention can slow down the ingestion speed. Configuring the [YB-TServers](../../reference/configuration/yb-tserver/) with multiple disks can reduce disk write contention, thereby increasing throughput. Disks with higher IOPS and better throughput also improve write performance.

- **Enabling packed columns** helps in increasing the throughput to more than two times. Enable packed columns on the YugabyteDB cluster by setting the YB-TServer gflag `ysql_enable_packed_row` to true.

> - **Client machine's disks** with higher IOPS and better throughput also improve import performance as a splitter, which splits the large data file into smaller splits of 20000 rows depending on the client machine's disk.

Reviewer: Client machine's disk with higher IOPS and better throughput also improves the import performance as, the performance of splitter which splits the large data file into smaller splits of 20000 rows depends on the client machine's disk.

yb-voyager maximizes performance when migrating data into a newly created empty database in several ways:

> - Creates secondary indexes after the data import is complete. During the [post-import data](../migrate-steps/#import-indexes-and-triggers) phase, yb-voyager creates the secondary indexes after it completes data loading on all the tables except the unique indexes which are created in the [import-schema](../migrate-steps/#import-schema) phase, only to avoid any foreign key dependencies on the index. Then, it uses [Index Backfill](https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/online-index-backfill.md) to update indexes after data is loaded. This is much faster than online index maintenance.

Reviewer 2: ```suggestion
- Creates secondary indexes after the data import is complete. During the [post-import data](../migrate-steps/#import-indexes-and-triggers) phase, yb-voyager creates the secondary indexes (except the unique indexes) after it completes data loading on all the tables. Then, it uses [Index Backfill](https://github.com/yugabyte/yugabyte-db/blob/master/architecture/design/online-index-backfill.md) to update indexes after data is loaded. This is much faster than online index maintenance. (Unique indexes are created in the [import-schema](../migrate-steps/#import-schema) phase, to avoid any foreign key dependencies on the index.)
```

> - Disables constraint checks except check constraints on the column. It's safe to disable all the constraint checks provided the data is from a reliable source. For maximum throughput, it is also preferable to not follow any order when populating tables. During the data import phase, yb-voyager disables all constraints, except for primary key violation errors.

Reviewer 3: This paragraph is a bit confusing in the present state.
How about the following paragraph:

> Disables foreign key constraints during import of data. However other constraints like primary key constraints, check constraints, unique key constraints etc are not disabled. It is safe to disable some constraint checks as the data is from a reliable source. For maximum throughput, it is also preferable to not follow any order when populating tables.

Author: Yes , calling out specifically about them is clearer.

  {{< note title="Note" >}}

> yb-voyager only disables all the constraints except check constraints on the column in internal sessions to migrate data.

Reviewer 2: ```suggestion
yb-voyager only disables constraint checks (except check constraints on columns) in internal sessions to migrate data.
```

> yb-voyager only disables constraint checks (except check constraints on columns) in internal sessions to migrate data.

Reviewer 3: Please take this _note_ section below the _Disables triggers_ point and modify the note content like this.

> yb-voyager only disables the constraint checks and triggers in the internal sessions it uses to migrate data.

Author: Sorry missed seeing this one and the top one. Thanks for pointing it out.

  {{< /note >}}

> - Disables triggers during import in the target database.

Reviewer 3: Replace with:
Disables triggers also during the import data phase.

### Techniques to improve performance

> Use one or more of the following techniques to improve import performance:

Reviewer 3: Let's make the "import data" explicit here.

> Use one or more of the following techniques to improve import data performance:

> - **Load data in parallel**. yb-voyager executes N parallel batch ingestion jobs at any given time, where N is equal to half of the total number of cores in the YugabyteDB cluster. Normally this is considered good default practice. However, if the target cluster runs on high resource machines with a large number of CPU cores, the default may result in underusing CPU resources.

Reviewer 3: **Existing**:

> Load data in parallel. yb-voyager executes N parallel batch ingestion jobs at any given time, where N is equal to half of the total number of cores in the YugabyteDB cluster. Normally this is considered good default practice. However, if the target cluster runs on high resource machines with a large number of CPU cores, the default may result in underusing CPU resources.

> Use the [-–parallel-jobs](https://deploy-preview-15688--infallible-bardeen-164bc9.netlify.app/preview/migrate/yb-voyager-cli/#parallel-jobs) argument with the import data command to override the default setting. Set --parallel-jobs to a number higher than the default based on your cluster configuration.

> If CPU use is greater than 50-60%, you should lower the number of jobs. Similarly, if CPU use is low, you can increase the number of jobs.

**Proposed**

> Load data in parallel. yb-voyager executes N parallel batch ingestion jobs at any given time, where N is equal to half of the total number of cores in the YugabyteDB cluster. Normally this is a good default value and should consume around 50-60 % cpu. However if the target cluster shows high cpu utilisation then one must lower the number of parallel jobs. Similarly if the target cluster seems to be under-utilised then it is recommended to stop the import and restart it with a higher number of parallel jobs. 

> Use the [-–parallel-jobs](https://deploy-preview-15688--infallible-bardeen-164bc9.netlify.app/preview/migrate/yb-voyager-cli/#parallel-jobs) argument with the import data command to override the default setting. Set --parallel-jobs to appropriate value to override the default based on your cluster configuration and observation.

>   Use the [-–parallel-jobs](../yb-voyager-cli/#parallel-jobs) argument with the import data command to override the default setting. Set `--parallel-jobs` to a number higher than the default based on your cluster requirement.

Reviewer 2: ```suggestion
  Use the [-–parallel-jobs](../yb-voyager-cli/#parallel-jobs) argument with the import data command to override the default setting. Set `--parallel-jobs` to a number higher than the default based on your cluster configuration.
```

  If CPU use is greater than 50-60%, you should lower the number of jobs. Similarly, if CPU use is low, you can increase the number of jobs.

   {{< note title="Note" >}}

>    If timeouts are witnessed, then import should be restarted with less parallelism.

Reviewer 2: ```suggestion
   If there are timeouts during import, restart the import with fewer parallel jobs.
```

   {{< /note >}}

> - **Increase batch size**. If the [--batch-size](../yb-voyager-cli/#batch-size) (default is 20000) is too small, the import will run slower because the time spent importing data may be comparable or less than the time spent on other tasks, such as bookkeeping, setting up the client connection, and so on. Increasing the batch-size to a very high value is not recommended as the whole batch is executed in one transaction.

Reviewer 3: **Existing**

> Increase batch size. If the [--batch-size](https://deploy-preview-15688--infallible-bardeen-164bc9.netlify.app/preview/migrate/yb-voyager-cli/#batch-size) (default is 20000) is too small, the import will run slower because the time spent importing data may be comparable or less than the time spent on other tasks, such as bookkeeping, setting up the client connection, and so on. Increasing the batch-size to a very high value is not recommended as the whole batch is executed in one transaction.

**Proposed**

> Increase batch size. The default [--batch-size](https://deploy-preview-15688--infallible-bardeen-164bc9.netlify.app/preview/migrate/yb-voyager-cli/#batch-size) is 20000 or approximately 200 MB of data depending on whichever is reached first while preparing the batch. Normally this is a good default value. However if the rows are too small then one can think of increasing the batch size for greater throughput. Increasing the batch-size to a very high value is not recommended as the whole batch is executed in one transaction.

- **Add disks** to reduce disk write contention. YugabyteDB servers can be configured with one or multiple disk volumes to store tablet data. If all tablets are writing to a single disk, write contention can slow down the ingestion speed. Configuring the [YB-TServers](../../reference/configuration/yb-tserver/) with multiple disks can reduce disk write contention, thereby increasing throughput. Disks with higher IOPS and better throughput also improve write performance.

> - **Enabling packed columns** helps in increasing the throughput to more than two times. Enable packed columns on the YugabyteDB cluster by setting the YB-TServer gflag `ysql_enable_packed_row` to true.

Reviewer 2: ```suggestion
- **Enable packed columns** to increase the throughput by more than two times. Enable packed columns on the YugabyteDB cluster by setting the YB-TServer gflag `ysql_enable_packed_row` to true.
```

> - **Client machine's disk** with higher IOPS and better throughput also improves the import performance as the performance of splitter which splits the large data file into smaller splits of 20000 rows depends on the client machine's disk.

Reviewer 2: ```suggestion
- **Configure the client [what is this, the source? target?] machine's disk** with higher IOPS and better throughput to improve the performance of the splitter, which splits the large data file into smaller splits of 20000 rows. Splitter performance depends on the client machine's disk.
```

Author: @priyanshi-yb - It is the target machine here ?

Reviewer: No, the machine on which Voyager is running.

Author: In that case , I'm going to go with host machine - https://docs.yugabyte.com/preview/migrate/install-yb-voyager/#prepare-the-host

    security_balance bigint DEFAULT 0 NOT NULL,
    security_nonce bigint DEFAULT 0 NOT NULL,
    balance bigint DEFAULT 0 NOT NULL,
    nonce bigint DEFAULT 0 NOT NULL,
    staked_balance bigint,
    PRIMARY KEY (block, address)
);
```

As more optimizations are introduced, average throughput increases. The following table shows the results.

| Run | Cluster configuration | yb-voyager flags | CPU usage | Average throughput |
| :-- | :-------------------- | :--------------- | :-------- | :----------------- |
| 24 parallel jobs (default) | 3 node [RF](../../architecture/docdb-replication/replication/#replication-factor) 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=24 | ~80% | 44014 rows/sec |
| Increase jobs<br>(1 per core) | 3 node RF 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48 | ~95% | 47696 rows/sec |
| Add nodes | 6 Node RF 3 cluster, c5.4x large (16 cores 32GB) <br> 4 EBS Type gp3 disks per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48 | ~80% | 86547 rows/sec |

> | Enabling packed columns | 3 node RF 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48 | ~95% | 134048 rows/sec |

Reviewer: in cluster config for this run, add GFLAG as well 
`GFLAG: ysql_enable_packed_row = true`

## Test results

yb-voyager was tested using varying configurations, including more parallel jobs, multiple disks, and a larger cluster. The tests were run using a 28GB CSV file with 350 million rows on YugabyteDB version 2.16.0.0-b90.

The table schema for the test was as follows:

```sql
CREATE TABLE public.accounts (

>     block bigint NOT NULL,discr

Reviewer 3: `discr` is spurious. Please remove.

    address text NOT NULL,
    dc_balance bigint DEFAULT 0 NOT NULL,
    dc_nonce bigint DEFAULT 0 NOT NULL,
    security_balance bigint DEFAULT 0 NOT NULL,
    security_nonce bigint DEFAULT 0 NOT NULL,
    balance bigint DEFAULT 0 NOT NULL,
    nonce bigint DEFAULT 0 NOT NULL,
    staked_balance bigint,
    PRIMARY KEY (block, address)
);
```

As more optimizations are introduced, average throughput increases. The following table shows the results.

| Run | Cluster configuration | yb-voyager flags | CPU usage | Average throughput |
| :-- | :-------------------- | :--------------- | :-------- | :----------------- |

> | 24 parallel jobs (default) | 3 node [RF](../../architecture/docdb-replication/replication/#replication-factor) 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=24 | ~80% | 44014 rows/sec |

Reviewer 2: maybe put each item in configuration column on separate line

cpu usage is all greater than 80%, yet above says don't go above 50-60?

Author: @ddhodge - I think it only means that we want the user to reduce the number of jobs if cpu usage goes above 50-60. I wonder if we are actually implying nit to go higher . But @priyanshi-yb , maybe you can tell us about it.

Reviewer: Yes, we don't recommend to higher than 50-60% and if it goes higher we recommend to reduce the parallel jobs. These results are for test purpose only to give users a ballpark idea of what perf numbers are there with different configurations.

| Increase jobs<br>(1 per core) | 3 node RF 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48 | ~95% | 47696 rows/sec |
| Add nodes | 6 Node RF 3 cluster, c5.4x large (16 cores 32GB) <br> 4 EBS Type gp3 disks per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48 | ~80% | 86547 rows/sec |

> | Enabling packed columns | 3 node RF 3 cluster, c5.4x large (16 cores 32 GB) <br> 1 EBS Type gp3 disk per node, 10000 IOPS, 500 MiB bandwidth | batch-size=20k<br>parallel-jobs=48<br>YB-TServer gflag: `ysql_enable_packed_row` = `true` | ~95% | 134048 rows/sec |

Reviewer 3: After this point we can add a note.

Note

> With increased optimisations the cpu % also shoots up. Letting the cpu go beyond 60 % may not be advisable as the database needs some cycles to do it's internal operations.
